### ISSUE TITLE
Nikon Z-series: The exif-tags from the raw file does not carry over to the exported jpeg file

### ISSUE THREAD

**Issue by purew** [Beep6581/RawTherapee#5698]
# Version details
```
Version: 5.8
Branch: releases
Commit: 9a9e0acbf
Commit date: 2020-02-04
Compiler: cc 9.2.0
Processor: x86_64
System: Linux
Bit depth: 64 bits
Gtkmm: V3.24.2
Lensfun: V0.3.95.0
Build type: Release
Build flags: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -std=c++11  -Werror=unused-label -Werror=delete-incomplete -Wall -Wuninitialized -Wcast-qual -Wno-deprecated-declarations -Wno-unused-result -fopenmp -Werror=unknown-pragmas -O3 -DNDEBUG -ftree-vectorize
Link flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
OpenMP support: ON
MMAP support: ON
```
# Operating system
```
Linux falcon 5.5.11-arch1-1 #1 SMP PREEMPT Sun, 22 Mar 2020 16:33:15 +0000 x86_64 GNU/Linux
```

# Problem

The exif-tags from the raw file does not carry over to the exported jpeg file

![tmp_007](https://user-images.githubusercontent.com/85910/78307114-33140580-7513-11ea-91eb-c7ff6bf07893.png)

Exported jpeg:
![2020-04-02_18-22-42_nikon z 6_dsc_0941](https://user-images.githubusercontent.com/85910/78307319-9ef66e00-7513-11ea-8778-1ca924a598e0.jpg)

**Comment by Nippius** [comment]
Hi, I'm having the same issue. 

Here's a couple of RAW files to show the problem (includes .pp3 and exported jpgs with missing tags:

https://filebin.net/kbdc0k317g9np2t9

## Version info:

### Equipment
Nikon Z50

### RawTherapee
Version: 5.8
Branch: releases
Commit: 9a9e0acbf
Commit date: 2020-02-04
Compiler: x86_64-w64-mingw32-gcc\ 8.2.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.24.2
Lensfun: V0.3.2.0
Build type: Release
Build flags: -mwin32\ -m64\ -mthreads\ -msse2\ -std=c++11\ -mtune=generic\ -Werror=unused-label\ -Werror=delete-incomplete\ -Wall\ -Wuninitialized\ -Wcast-qual\ -Wno-deprecated-declarations\ -Wno-unused-result\ -fopenmp\ -Werror=unknown-pragmas\ -Wno-aggressive-loop-optimizations\ -DNDEBUG\ -O3\ -ftree-vectorize
Link flags: -m64\ -mthreads\ -static-libgcc\ -mtune=generic\ -s\ -O3
OpenMP support: ON
MMAP support: ON

### Windows
![image](https://user-images.githubusercontent.com/1528640/78828834-3c064a80-79dd-11ea-8836-2530f69b01c2.png)
(x64 version)

**Comment by purew** [comment]
> ### Equipment
> 
> Nikon Z50

I'm using the Nikon Z6. Seeing that you are using Z 50 made me wonder if this is due to some change in the Z-series. So I tried exporting a raw from my older Nikon d7500 and that worked fine.

TLDR: Seems like this may be due to a change in the Z-series...

**Comment by Nippius** [comment]
That's a good find! Let me try with my D3500 to see if it works.

**Comment by Nippius** [comment]
Yup raw files from my d3500 are exported correctly

**Comment by purew** [comment]
I'd be happy digging into this and submitting a patch. Any pointers as to where I should start digging?

**Comment by heckflosse** [comment]
@PureW Great that you want to help :+1: 
Maybe start here:
https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/imageio.cc#L1230

**Comment by purew** [comment]
> @PureW Great that you want to help +1
> Maybe start here:
> https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/imageio.cc#L1230

That was simple, the size is 71924 which is more than 65530 and the write wont be triggered at https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/imageio.cc#L1233

I'm not sure though why the jpeg is not allowed to be larger and how this should be resolved?

**Comment by heckflosse** [comment]
> I'm not sure though why the jpeg is not allowed to be larger and how this should be resolved?

Maybe there's a large block in exif data we can exclude for the jpeg?

**Comment by Thanatomanic** [comment]
Wikipedia has this to say about the size limit: https://en.wikipedia.org/wiki/Exif#Technical_2
(fifth point)

**Comment by purew** [comment]
I'm printing the exif tags as seen by rawtherapee during export and `ShotInfo: size 53670` and `ApplicationNotes: size 12288` stands out:


```
In createJPEGMarker              
is root                     
- NewSubFileType: size 4         
- ImageWidth: size 4        
- ImageHeight: size 4           
- BitsPerSample: size 6       
- Compression: size 2         
- PhotometricInterpretation: size 2
- Make: size 18                        
- Model: size 10             
- Orientation: size 2    
- SamplesPerPixel: size 2       
- XResolution: size 8            
- YResolution: size 8       
- PlanarConfiguration: size 2
- ResolutionUnit: size 2
- Software: size 13           
- DateTime: size 20           
- Artist: size 37         
+-- DIRECTORY SubIFD[0]:  
    - NewSubFileType: size 4       
    - Compression: size 2     
    - XResolution: size 8    
    - YResolution: size 8    
    - ResolutionUnit: size 2          
    - JPEGInterchangeFormat: size 4
    - JPEGInterchangeFormatLength: size 4
    - YCbCrPositioning: size 2 
+-- DIRECTORY SubIFD[1]:  
    - NewSubFileType: size 4
    - ImageWidth: size 4
    - ImageHeight: size 4      
    - BitsPerSample: size 2    
    - Compression: size 2         
    - PhotometricInterpretation: size 2
    - SamplesPerPixel: size 2  
    - XResolution: size 8  
    - YResolution: size 8      
    - PlanarConfiguration: size 2
    - ResolutionUnit: size 2          
    - CFAPatternDim: size 4     
    - CFAPattern: size 4      
+-- DIRECTORY SubIFD[2]:            
    - NewSubFileType: size 4               
    - Compression: size 2             
    - XResolution: size 8     
    - YResolution: size 8     
    - ResolutionUnit: size 2
    - JPEGInterchangeFormat: size 4
    - JPEGInterchangeFormatLength: size 4
    - YCbCrPositioning: size 2
    - JPEGInterchangeFormatLength: size 4                                                                                                                                                                                             [61/3950]
    - YCbCrPositioning: size 2                                                                                         
+-- DIRECTORY SubIFD[3]:                                                                                               
    - NewSubFileType: size 4                                                                                           
    - ImageWidth: size 4                                                                                               
    - ImageHeight: size 4                                                                                              
    - BitsPerSample: size 4                                                                                            
    - Compression: size 2                                                                                              
    - PhotometricInterpretation: size 2                                                                                
    - SamplesPerPixel: size 2                                                                                          
    - XResolution: size 8                                                                                              
    - YResolution: size 8                             
    - PlanarConfiguration: size 2
    - ResolutionUnit: size 2
+-- DIRECTORY SubIFD[4]:         
    - NewSubFileType: size 4
    - ImageWidth: size 4        
    - ImageHeight: size 4     
    - BitsPerSample: size 2   
    - Compression: size 2          
    - PhotometricInterpretation: size 2
    - SamplesPerPixel: size 2
    - XResolution: size 8
    - YResolution: size 8       
    - PlanarConfiguration: size 2
    - ResolutionUnit: size 2
    - CFAPatternDim: size 4  
    - CFAPattern: size 4
- ReferenceBlackWhite: size 48
- ApplicationNotes: size 12288
- Copyright: size 55      
+-- DIRECTORY Exif[0]:    
    - ExposureTime: size 8         
    - FNumber: size 8         
    - ExposureProgram: size 2
    - ISOSpeedRatings: size 2
    - RecommendedExposureIndex: size 4
    - DateTimeOriginal: size 20    
    - DateTimeDigitized: size 20         
    - ExposureBiasValue: size 8
    - MeteringMode: size 2
    - LightSource: size 2   
    - Flash: size 2     
    - FocalLength: size 8      
    +-- DIRECTORY MakerNote[0]:
        - MakerNoteVersion: size 4
        - Quality: size 8              
        - WhiteBalance: size 13
        - FocusMode: size 7
        - FlashSetting: size 13
        - FlashType: size 20     
        - WhiteBalanceFineTune: size 4
        - ColorBalance1: size 32
        - ProgramShift: size 4
        - ExposureDifference: size 4
        - FlashExposureBracketValue: size 4
        - ExposureBracketValue: size 8
        - CropHiSpeed: size 14
        - SerialNumber: size 8
        - ColorSpace: size 2
        +-- DIRECTORY ISOInfo[0]:  
            - ISO: size 1                
            - ISOExpansion: size 2
            - ISO: size 1                                                                                                                                                                                                              [0/3950]
            - ISOExpansion: size 2                                                                                     
            - ISO2: size 1                                                                                             
            - ISOExpansion2: size 2                                                                                    
        - LensType: size 1                                                                                             
        - Lens: size 32                                                                                                
        - FlashMode: size 1                                                                                            
        - ShootingMode: size 2                                                                                         
        - AutoBracketRelease: size 2                                                                                   
        - LensFStops: size 4                                                                                           
        - NEFCurve1: size 578                                                                                          
        - ShotInfo: size 53670                        
        - NoiseReduction: size 5 
        - NEFCurve2: size 46
        - ColorBalance: size 3316
        - LensData: size 108
        - RawImageCenter: size 4
        - ShutterCount: size 4
        - VariProgram: size 16
        - MultiExposure: size 16   
        - HighISONoiseReduction: size 2
    - UserComment: size 44   
    - SubSecTime: size 3 
    - SubSecTimeOriginal: size 3
    - SubSecTimeDigitized: size 3
    - SensingMethod: size 2 
    - FileSource: size 1     
    - SceneType: size 1 
    - CFAPattern: size 8      
    - CustomRendered: size 2  
    - ExposureMode: size 2
    - WhiteBalance: size 2
    - FocalLengthIn35mmFilm: size 2
    - SceneCaptureType: size 2
    - GainControl: size 2    
    - Contrast: size 2       
    - Saturation: size 2              
    - Sharpness: size 2            
    - SubjectDistanceRange: size 2       
    - SerialNumber: size 8     
    - LensInfo: size 32   
    - LensMake: size 6      
    - LensModel: size 65
    - LensSerialNumber: size 11
+-- DIRECTORY GPSInfo[0]:      
    - GPSVersionID: size 4        
- DateTimeOriginal: size 20            
- Rating: size 4
```

For reference, here's the exifdump from a camera jpeg:
```bash
$ exiftool 2020-03-28_12-11-43_nikon\ z\ 6_dsc_0735.jpg                                                                                                                                               [152/389]
ExifTool Version Number         : 12.00  
File Name                       : 2020-03-28_12-11-43_nikon z 6_dsc_0735.jpg
Directory                       : .                  
File Size                       : 6.8 MB    
File Modification Date/Time     : 2020:03:28 12:11:44-04:00 
File Access Date/Time           : 2020:03:28 12:11:44-04:00                                                            
File Inode Change Date/Time     : 2020:03:28 12:20:36-04:00 
File Permissions                : rw-r--r--
File Type                       : JPEG            
File Type Extension             : jpg
MIME Type                       : image/jpeg
Exif Byte Order                 : Little-endian (Intel, II) 
Make                            : NIKON CORPORATION
Camera Model Name               : NIKON Z 6
Orientation                     : Horizontal (normal)
X Resolution                    : 300  
Y Resolution                    : 300                                                                                  
Resolution Unit                 : inches                                                                               
Software                        : Ver.03.00                                                                            
Modify Date                     : 2020:03:28 12:11:43                                                                  
Artist                          : ANDERS BENNEHAG                                                                      
Y Cb Cr Positioning             : Co-sited                                                                             
Copyright                       :     
Exposure Time                   : 1/160         
F Number                        : 4.0     
Exposure Program                : Manual                
ISO                             : 1250             
Sensitivity Type                : Recommended Exposure Index                                                           
Recommended Exposure Index      : 1250   
Exif Version                    : 0231                                                                                 
Date/Time Original              : 2020:03:28 12:11:43
Create Date                     : 2020:03:28 12:11:43
Offset Time                     : -04:00
Offset Time Original            : -04:00
Offset Time Digitized           : -04:00
Components Configuration        : Y, Cb, Cr, -
Compressed Bits Per Pixel       : 4
Exposure Compensation           : 0
Metering Mode                   : Multi-segment
Light Source                    : Unknown
Flash                           : No Flash
Focal Length                    : 70.0 mm
Maker Note Version              : 2.11
Quality                         : Fine
White Balance                   : Auto1
Focus Mode                      : AF-C
Flash Setting                   : 
Flash Type                      : 
White Balance Fine Tune         : 0 0
WB RB Levels                    : 1.8203125 1.24609375 1 1
Program Shift                   : 0
Exposure Difference             : -1.6
Image Boundary                  : 0 0 6048 4024
External Flash Exposure Comp    : 0
Flash Exposure Bracket Value    : 0.0
Exposure Bracket Value          : 0
Crop Hi Speed                   : FX Uncropped (6064x4040 cropped to 6064x4040 at pixel 0,0)
Exposure Tuning                 : 0
VR Info Version                 : 0200
Vibration Reduction             : On
VR Mode                         : On (1)
Active D-Lighting               : Off
Picture Control Version         : 0300                                                                                 
Picture Control Name            : Monochrome         
Picture Control Base            : Monochrome
Picture Control Adjust          : Quick Adjust                                                                         
Picture Control Quick Adjust    : n/a                                                                                  
Mid Range Sharpness             : 2.00                                                                                 
Clarity                         : 1.00     
Brightness                      : Normal          
Hue                             : n/a
Filter Effect                   : Off       
Toning Effect                   : B&W                                                                                  
Toning Saturation               : n/a              
Time Zone                       : -05:00   
Daylight Savings                : Yes                
Date Display Format             : Y/M/D
ISO Expansion                   : Off                                                                                  
ISO2                            : 1270                                                                                 
ISO Expansion 2                 : Off                                                                                  
Vignette Control                : Normal                                                                               
Auto Distortion Control         : On                                                                                   
Shutter Mode                    : Unknown (81)                                                                         
HDR Info Version                : 0100
HDR                             : Off           
HDR Level                       : Auto    
HDR Smoothing                   : Off                   
HDR Level 2                     : n/a              
Mechanical Shutter Count        : 671                                                                                  
Color Temperature Auto          : 5780   
Lens Type                       : AF                                                                                   
Lens                            : 24-70mm f/4        
Flash Mode                      : Did Not Fire       
Shooting Mode                   : Continuous    
Lens F Stops                    : 5.00                 
Shot Info Version               : 0800                   
Noise Reduction                 : Off         
Color Balance Version           : 0800  
Lens Data Version               : 0800 
Max Aperture                    : 4.0          
Focus Distance                  : 2.82 m     
Retouch History                 : None    
Image Data Size                 : 6372423
Shutter Count                   : 733   
Flash Info Version              : 0108   
Flash Source                    : None   
External Flash Firmware         : n/a        
External Flash Zoom Override    : No   
External Flash Status           : Flash Not Attached    
External Flash Ready State      : n/a     
Flash Compensation              : 0                       
Flash GN Distance               : 0               
Flash Group A Control Mode      : Off  
Flash Group B Control Mode      : Off          
Flash Group C Control Mode      : Off                
Flash Group A Compensation      : 0  
Flash Group B Compensation      : 0   
Flash Group C Compensation      : 0                                                                                    
Vari Program                    :                      
Multi Exposure Version          : 0101
Multi Exposure Mode             : Off                                                                                  
Multi Exposure Shots            : 0     
Multi Exposure Auto Gain        : Off
Multi Exposure Shots            : 0                                                                                                                                                                                                    [30/389]
Multi Exposure Auto Gain        : Off    
High ISO Noise Reduction        : Normal                                                                               
Power Up Time                   : 2020:03:23 18:20:15
AF Info 2 Version               : 0300      
Contrast Detect AF              : On (2)                                                                               
AF Area Mode                    : Auto                                                                                 
Phase Detect AF                 : On (81-point)                                                                        
Primary AF Point                : (none)   
AF Points Used                  : B4,C4,C5,D4,D5  
AF Area X Position              : 0  
AF Area Y Position              : 0         
File Info Version               : 0100                                                                                 
Memory Card Number              : 0                
Directory Number                : 100      
File Number                     : 0735               
AF Fine Tune                    : Off  
AF Fine Tune Index              : n/a                                                                                  
AF Fine Tune Adj                : 0                                                                                    
AF Fine Tune Adj Tele           : 0                                                                                    
Retouch Info Version            : 0200                                                                                 
Retouch NEF Processing          : Off                                                                                  
User Comment                    :                                                                                      
Sub Sec Time                    : 91  
Sub Sec Time Original           : 91            
Sub Sec Time Digitized          : 91      
Flashpix Version                : 0100                  
Color Space                     : sRGB             
Exif Image Width                : 6048                                                                                 
Exif Image Height               : 4024   
Interoperability Index          : R98 - DCF basic file (sRGB)                                                          
Interoperability Version        : 0100               
Sensing Method                  : One-chip color area
File Source                     : Digital Camera
Scene Type                      : Directly photographed
CFA Pattern                     : [Red,Green][Green,Blue]
Custom Rendered                 : Custom      
Exposure Mode                   : Manual
Focal Length In 35mm Format     : 70 mm
Scene Capture Type              : Standard     
Gain Control                    : Low gain up
Contrast                        : Normal  
Saturation                      : Normal 
Sharpness                       : Normal
Subject Distance Range          : Unknown
Serial Number                   : 3054814
Lens Info                       : 24-70mm f/4
Lens Make                       : NIKON
Lens Model                      : NIKKOR Z 24-70mm f/4 S
Lens Serial Number              : 20112047
GPS Version ID                  : 2.3.0.0                 
Compression                     : JPEG (old-style)
Thumbnail Offset                : 32380
Thumbnail Length                : 8492         
Creator Tool                    : NIKON Z 6 Ver.03.00
Rating                          : 0  
MPF Version                     : 0100
Number Of Images                : 3                                                                                    
MP Image Flags                  : Dependent child image
MP Image Format                 : JPEG
MP Image Type                   : Large Thumbnail (full HD equivalent)                                                 
MP Image Length                 : 491379
MP Image Start                  : 6589440
Dependent Image 1 Entry Number  : 0             
Dependent Image 2 Entry Number  : 0                    
Image Width                     : 6048                   
Image Height                    : 4024        
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8    
Color Components                : 3            
Y Cb Cr Sub Sampling            : YCbCr4:2:2 (2 1)
Aperture                        : 4.0     
Blue Balance                    : 1.246094
Image Size                      : 6048x4024
Megapixels                      : 24.3   
Red Balance                     : 1.820313
Scale Factor To 35 mm Equivalent: 1.0        
Shutter Speed                   : 1/160
Create Date                     : 2020:03:28 12:11:43.91-04:00
Date/Time Original              : 2020:03:28 12:11:43.91-04:00
Modify Date                     : 2020:03:28 12:11:43.91-04:00
Thumbnail Image                 : (Binary data 8492 bytes, use -b option to extract)
Preview Image                   : (Binary data 132802 bytes, use -b option to extract)
MP Image 3                      : (Binary data 491379 bytes, use -b option to extract)
Auto Focus                      : On                 
Lens Spec                       : 24-70mm f/4 AF
Circle Of Confusion             : 0.030 mm
Depth Of Field                  : 0.38 m (2.64 - 3.02 m)                                                               
Field Of View                   : 28.2 deg (1.41 m)    
Focal Length                    : 70.0 mm (35 mm equivalent: 70.0 mm)
Hyperfocal Distance             : 40.77 m                                                                              
Light Value                     : 7.7
```

**Comment by heckflosse** [comment]
According to https://exiftool.org/TagNames/EXIF.html

ApplicationNotes notes are XMP tags (not from RawTherapee for sure). Maybe we should just exclude them from the jpeg

**Comment by purew** [comment]
> According to https://exiftool.org/TagNames/EXIF.html
> 
> ApplicationNotes notes are XMP tags (not from RawTherapee for sure). Maybe we should just exclude them from the jpeg

I opened a PR to drop just the ShotInfo. Would you like me to amend it to drop the entire `MakerNotes`?

**Comment by heckflosse** [comment]
@PureW 
> Would you like me to amend it to drop the entire MakerNotes?

Definitely not. A lot of cameras have some infos in MakerNote sub tag which might by good to keep

**Comment by heckflosse** [comment]
@PureW I will have a look now what's inside the MakerNotes.ShotInfo of Nikon Z-Series...

**Comment by heckflosse** [comment]
@PureW `MakerNotes.ShotInfo` Looks like some data which is not interpreted by RawTherapee at all
![grafik](https://user-images.githubusercontent.com/13733487/85232620-928dfd00-b400-11ea-9637-118d80ddc7dd.png)

Should be ok to not include it into jpeg exif data.
